document-portal/file-transfer.c

```diff
@@ -0,0 +1,447 @@
+
+  return transfer;
+}
+
+static FileTransfer *
+file_transfer_start (XdpAppInfo *app_info,
+                     const char *sender,
+                     gboolean    writable,
+                     gboolean    autostop)
+{
+  FileTransfer *transfer;
+
+  transfer = g_new0 (FileTransfer, 1);
+
+  g_mutex_init (&transfer->mutex);
+
+  transfer->app_info = xdp_app_info_ref (app_info);
+  transfer->sender = g_strdup (sender);
+  transfer->writable = writable;
+  transfer->autostop = autostop;
+  transfer->key = g_strdup_printf ("%u", g_random_int ());
+  transfer->files = g_ptr_array_new_with_free_func (g_free);
+
+  G_LOCK (transfers);
+  g_hash_table_insert (transfers, transfer->key, transfer);
+  G_UNLOCK (transfers);
+
+  g_debug ("start file transfer owned by '%s' (%s)",
+           xdp_app_info_get_id (transfer->app_info),
+           transfer->sender);
+
+  return transfer;
+}
+
+static gboolean
+stop (gpointer data)
+{
+  FileTransfer *transfer = data;
+
+  file_transfer_free (transfer);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+file_transfer_stop (FileTransfer *transfer)
+{
+  g_debug ("stop file transfer owned by '%s' (%s)",
+           xdp_app_info_get_id (transfer->app_info),
+           transfer->sender);
+
+  G_LOCK (transfers);
+  g_hash_table_steal (transfers, transfer->key);
+  G_UNLOCK (transfers);
+
+  g_idle_add (stop, transfer);
+}
+
+static void
+file_transfer_add_files (FileTransfer *transfer,
+                         const char **files)
+{
+  int i;
+
+  for (i = 0; files[i]; i++)
+    g_ptr_array_add (transfer->files, g_strdup (files[i]));
+
+  g_debug ("add %d files to file transfer owned by '%s' (%s)", g_strv_length (files),
+           xdp_app_info_get_id (transfer->app_info),
+           transfer->sender);
+}
+
+static char **
+file_transfer_execute (FileTransfer *transfer,
+                       XdpAppInfo *target_app_info,
+                       GError **error)
+{
+  guint32 flags;
+  DocumentPermissionFlags perms;
+  const char *target_app_id;
+  int n_fds;
+  g_autofree int *fds = NULL;
+  int i;
+  g_auto(GStrv) ids = NULL;
+  char **files = NULL;
+
+  g_debug ("retrieve %d files for %s from file transfer owned by '%s' (%s)",
+           transfer->files->len,
+           xdp_app_info_get_id (target_app_info),
+           xdp_app_info_get_id (transfer->app_info),
+           transfer->sender);
+
+  if (xdp_app_info_is_host (target_app_info))
+    return g_strdupv ((char **)transfer->files->pdata);
+
+  flags = DOCUMENT_ADD_FLAGS_REUSE_EXISTING | DOCUMENT_ADD_FLAGS_AS_NEEDED_BY_APP;
+  
+  perms = DOCUMENT_PERMISSION_FLAGS_READ;
+  if (transfer->writable)
+    perms |= DOCUMENT_PERMISSION_FLAGS_WRITE;
+
+  target_app_id = xdp_app_info_get_id (target_app_info);
+
+  n_fds = transfer->files->len;
+  fds = g_new (int, n_fds);
+  for (i = 0; i < n_fds; i++)
+    {
+      const char *file;
+
+      file = g_ptr_array_index (transfer->files, i);
+      fds[i] = open (file, O_PATH | O_CLOEXEC);
```

Comment:
While this works, its kinda weird to have the portal open the paths (which we already verified the calling app could access) and pass the FDs into document_add_full(). Maybe we could have a version of document_add that takes paths?

Reply:
but I think we need fds in document_add_full anyway, to check if the target can access them ?